src/twisted/internet/testing.py

```diff
@@ -669,8 +669,16 @@ def callWhenRunning(
         """
         Fake L{IReactorCore.callWhenRunning}.
         Keeps a list of invocations to make in C{self.whenRunningHooks}.
+
+        If the reactor has not started, the callable will be scheduled
+        to run when it does start. Otherwise, the callable will be invoked
+        immediately.
         """
-        self.whenRunningHooks.append((callable, args, kw))
+        if self.running:
+            callable(*args, **kw)
+            return None
```

Comment:
Should we additionally append to `self.whenRunningHooks` in this case?

This would match the previous behavior but the name `whenRunningHooks` describes a list that only includes hooks that should be invoked *when* the reactor is running. Whereas, these are callbacks *after* the reactor is running.

I'm guessing we may want to always append to `self.whenRunningHooks` to avoid breaking peoples tests that rely on this behavior 🤷

Reply:
I think that `whenRunningHooks` should have been private and people should not depend on it.

Also... when `reactor.run()` is called, I think that `whenRunningHooks` should be cleared...

I don't know what to say.

My call is to go with this implementation...and we can see if people will complain